### ISSUE TITLE
CHARMM and AMBER forcefield updates

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1611]
Following our conversation with @peastman, @rafwiewiora, and @ChayaSt last Friday, I've created a new github repo to hold the forcefield conversion code, source forcefields, automated testing harness, and resulting `ffxml` files:

https://github.com/choderalab/openmm-forcefields

Once this has reached stability, we will transfer it to the `pandegroup` github org.

I'll draft a [project board](https://github.com/choderalab/openmm-forcefields/projects) shortly, but I just wanted to make sure I list all of the essential tasks:
- migrate all conversion code and input forcefields for CHARMM
- migrate all conversion code and input forcefields for AMBER
- add capability to convert CHARMM patches to new `ffxml` format to ParmEd
- extend `ffxml` format to allow inclusion of multiple `ffxml` files in a metapackage
- identify metapackage `ffxml`s to be built for AMBER; CHARMM will be c36
- identify whether additional CHARMM forcefields need to be converted
- finalize and implement solution for GBSA that works for new protein and small molecule `ffxml` formats
- update [`ambermini`](https://github.com/choderalab/ambermini) to AmberTools16
- migrate [GAFF forcefield generators](https://github.com/choderalab/openmoltools/blob/master/openmoltools/forcefield_generators.py) to allow small molecule support
- automate conversion and testing with driver scripts
- automate conversion/testing via travis, build conda dev package
- add documentation
- build conda recipe for optional `openmm-forcefields` package containing all forcefields
- copy converted `ffxml` files we will use in OpenMM into OpenMM release with updated forcefields

I estimate this will take 1-2 months, so recommend we proceed with OpenMM 7.1 supporting CUDA 8.0 and cut a 7.2 once this is ready.

**Comment by peastman** [comment]
Could we get a new time estimate on this?  It's been four months since you estimated one to two months, so clearly that isn't correct.  Also, this is a prerequisite for being able to implement https://github.com/pandegroup/pdbfixer/issues/144, so it's now blocking other features.

**Comment by rafwiewiora** [comment]
I have working on finishing all Amber FF parts planned for during the BPS conference, starting on Friday. Won't take longer than a few days + time for discussions.

**Comment by peastman** [comment]
Great, thanks!

**Comment by peastman** [comment]
What's the status of this?  It's been another three weeks since you said it would take only a few days.

**Comment by peastman** [comment]
We really really *really* need to get this done **right now**.  To give a sense of just how far behind schedule this is, on Dec. 3, 2015 @jchodera described the Amber and CHARMM force field conversions as "pretty close" to done, and said he was "confident" they would be finished in a few weeks (#1282).  That was in discussing the feature list for 7.0.  But it wasn't done in time for 7.0, and then it wasn't done in time for 7.1, and now it's blocking me from working on core features of 7.2.  I have no clue what's been going on all that time, except that clearly no one has been doing the things they repeatedly said they were going to do.

@rafwiewiora and @ChayaSt can you have this done by Friday?

**Comment by rafwiewiora** [comment]
Alright, uploading stuff to Github - PR for impropers code https://github.com/pandegroup/openmm/pull/1560 ready to merge in a sec, next stuff flowing into https://github.com/choderalab/openmm-forcefields.

**Comment by jchodera** [comment]
We won't have this done by Friday, @peastman. There's a lot of essential tasks in https://github.com/pandegroup/openmm/issues/1611#issue-181292335 and there's no way we can get all of that done by then. This will take at least a month to complete in total.

**What specific features do you need immediately?** We may be able to deliver the specific functionality you need right away. I understand that https://github.com/pandegroup/pdbfixer/issues/144 is waiting on the availability of ffxml versions of lipid forcefields, and @rafwiewiora is supposed to see if the AmberTools16 lipid forcefields may meet your needs (which could be made available/installable immediately), and @ChayaSt has ported her PR https://github.com/choderalab/openmm-forcefields/pull/1 that has some CHARMM features available immediately (without some patches converted that aren't currently supported by ParmEd). If you can help us determine precisely what is needed, we can likely get that installable as soon as the Omnia build infrastructure fully recovers from yesterday's S3 outage and the python 3.6 additions.

**Comment by jchodera** [comment]
@peastman : From https://github.com/pandegroup/pdbfixer/issues/144#issuecomment-272015898, it looks like you wanted:
* A summary of which lipids are supported by the Amber and CHARMM forcefields we're adding in 7.2
* Access to the `ffxml` files for those lipids

Is that right?

* @rafwiewiora and @ChayaSt : How hard would it be to compile the list of available lipids? 
* @peastman : How would you want us to summarize the available lipids?
* @peastman : What if we just made *some* lipids available immediately, and then expanded this to the whole list later? Would that be sufficient?

**Comment by rafwiewiora** [comment]
I'll see if I can make at least one AMBER lipid forcefield work, right after I get the PRs I wanted to finish today done.

**Comment by peastman** [comment]
If you can create a force field that includes even just POPC and cholesterol (plus protein and water, of course), that would be enough to let me start working on it.

**Comment by jchodera** [comment]
@lnaden: Can you also assist @rafwiewiora if needed since you updated AmberMini to v16?

**Comment by ChayaSt** [comment]
Since many of CHARMM lipid forcefield files doesn't have any patches, I can convert those files by Friday. However, I haven't tested the conversion and I don't think I'll have time to do the full testing by Friday.

**Comment by peastman** [comment]
If you can just send me whatever you have, that would be great, even if it isn't fully tested yet.

**Comment by ChayaSt** [comment]
[lipid ffxml](https://github.com/ChayaSt/openmm-forcefields/blob/charmm36/charmm/ffxml/charmm36_lipid.xml)

[yaml file](https://github.com/ChayaSt/openmm-forcefields/blob/charmm36/charmm/charmm36_lipid.yaml) that specifies the files used to generate the lipid ffxml file. I excluded stream files that either had patches or that used atom types that are not in the main lipid force field. 

This has not been tested yet.

**Comment by peastman** [comment]
Awesome, thanks!

I'll need proteins as well.

**Comment by rafwiewiora** [comment]
@peastman lipid14 from Amber - ffxml here: https://github.com/choderalab/openmm/blob/forcefield_conversion_bonanza_amber16/devtools/forcefield-scripts/amber/ffxml/lipid14.xml

This should work no problem with current Amber protein ffxmls I think!

**Comment by jchodera** [comment]
> I'll need proteins as well.

You mean the charmm36 protein forcefields from @ChayaSt, right?

**Comment by peastman** [comment]
Thanks!

**Comment by peastman** [comment]
I'm running into issues with the lipids, and I'm not sure of the best way of handling them.  Mostly it seems to be a lack of standardization.

I found a couple of sites that have PDB files for pre-equilibrated membrane patches:

http://cmb.bio.uni-goettingen.de/cholmembranes.html
http://www.fos.su.se/~sasha/SLipids/Downloads.html

Each of them represents the lipids in the same way.  Each POPC molecule is a single residue with the nonstandard identifier POP, and just to make things more convenient, no CONECT records.  (sigh)  The CHARMM force field also represents them that way: a single residue called POPC.  The Amber force field, in contrast, represents each molecule as being made up of three residues: PA, OL, and PC (I think that's right?).

How do we handle this situation?

**Comment by peastman** [comment]
Oh, and just to make things *even more convenient*, the second of those two sites listed above uses exactly the same identifier (POP) for both POPC and POPE.  Because, you know, having different identifiers for different molecules just might confuse someone or something.

**Comment by swails** [comment]
Ambers way is nonstandard. Charmms way is standard. Amber utilizes a preprocessing script to break up the residues into segments that match the residue templates. 

It may be easiest to have the ff conversion script combine the templates necessary to create templates for the common lipid residues. You lose some of the flexibility doing that, but it seems like flexibility we wouldn't take advantage of anyway. 

Another alternative is to somehow incorporate the logic in charmmlipid2amber.py into OpenMM.

**Comment by swails** [comment]
Or charmm may not be standard, but Amber starts from CHARMM and then splits lipid residues into pieces to match the residue templates in Amber.

**Comment by peastman** [comment]
> It may be easiest to have the ff conversion script combine the templates necessary to create templates for the common lipid residues.

That sounds to me like it's probably the best approach.  All the PDB files I've found online use a single residue per lipid molecule, so that seems to be the standard.

**Comment by peastman** [comment]
Any updates on this?  At the moment I have a CHARMM force field for lipids but no proteins, and an AMBER force field that doesn't match the way lipids are defined in PDB files.  I hoped I could just merge the residues by hand, but it looks like that would be something of a task because the atom names aren't unique.

**Comment by peastman** [comment]
Any updates on this?  Anyone?  It's been over a month, so I assume you're getting close to having it finished?

**Comment by jchodera** [comment]
Sorry for the delay. I've been traveling, but will be back tomorrow and will give an update and timeline after I figure out where we are at the moment.

**Comment by jchodera** [comment]
I've corresponded with @rafwiewiora and updated the [`openmm-forcefields` project board](https://github.com/choderalab/openmm-forcefields/projects/1) with remaining tasks. We'll try to keep this up to date so you can see where things are. 

@peastman : Can you accept the invite to be a collaborator on that repo? We need to assign a few issues to you regarding changes that directly impact the https://github.com/pandegroup/openmm repo.

To recap, our strategy is to have https://github.com/choderalab/openmm-forcefields house all the forcefield conversion code, source parameters, final (converted) parameters, and forcefield generators. This will allow us to keep all the conversion code together, automate testing and validation, etc. 

We will build a separate `openmm-forcefields` conda package containing all of the AMBER and CHARMM forcefields, and users can install this optional package if they wish to get access to everything. We still have to figure out where this will install to, but once we add this path to the default search path for OpenMM, users should be able to access the recommended new forcefield bundles from the latest AmberTools package with
```python
ff = ForceField('amber/ff99SB14.xml`) 
```
or the individual ("a la carte") forcefields from a *specific* forcefield package (for reproducibility) with:
```python
ff = ForceField('amber/ambertools16/leaprc.protein.ff14SB.xml`, 'amber/ambertools16/parm10.dat.xml`) 
```
All old `ffxml` files will still be accessible in the same place in the same way for backwards-compatibility, 
```python
ff = ForceField('amber99sbildn.xml`) 
```
New `ffxml` files that @peastman wants to add to the top-level here can be cherry-picked and dropped into OpenMM's `data/` directory for distribution with 7.2. 

@peastman : Feel free to create corresponding OpenMM issues for the things in the OpenMM column on https://github.com/choderalab/openmm-forcefields/projects/1 if you think that's the best way to move forward with those changes that directly impact this repo.

**Comment by peastman** [comment]
Thanks!  Can we assign completion dates to everything?  It looks like the two things I need to do are https://github.com/choderalab/openmm-forcefields/issues/4 and https://github.com/choderalab/openmm-forcefields/issues/6.  I'll aim to have both of them done by the end of the day on Monday, hopefully sooner.

I didn't see any mention of the fixes to the Amber lipids discussed above.  Does that mean it's already been completed?  If so, can I get the updated force field?

More generally, there haven't been any commits to that repository in a month and a half, so I can't tell from it what anyone has been working on.  @rafwiewiora and @ChayaSt can you each post a summary of what you've been working on and which tasks have been completed since you last posted here?  Thanks!

**Comment by jchodera** [comment]
> Thanks! Can we assign completion dates to everything? 

I'll work on that tonight-tomorrow.

> It looks like the two things I need to do are choderalab/openmm-forcefields#4 and choderalab/openmm-forcefields#6. I'll aim to have both of them done by the end of the day on Monday, hopefully sooner.

I've assigned these items (and a couple more in the OpenMM card) to you. Feel free to punt on https://github.com/choderalab/openmm-forcefields/issues/5, and https://github.com/choderalab/openmm-forcefields/issues/21 can't happen until later.

> I didn't see any mention of the fixes to the Amber lipids discussed above. Does that mean it's already been completed? If so, can I get the updated force field?

Can you summarize or point to the issue that needs to be addressed, or add the appropriate issue to the projects board?

> More generally, there haven't been any commits to that repository in a month and a half, so I can't tell from it what anyone has been working on. @rafwiewiora and @ChayaSt can you each post a summary of what you've been working on and which tasks have been completed since you last posted here? Thanks!

@ChayaSt has migrated her files to the `openmm-forcefields` repo, and @rafwiewiora is working on doing this now. I think @ChayaSt provided you with some files you needed in the interim that had already been converted.

Once everyone is back at work next week, we should be in good shape to wrap this up quickly. Let's check in on Wednesday to see where things are and what remains to be done. A video chat might make it easier to sync up on any unresolved issues.

**Comment by peastman** [comment]
> Can you summarize or point to the issue that needs to be addressed

https://github.com/pandegroup/openmm/issues/1611#issuecomment-288887575 and the discussion following it.

> @ChayaSt has migrated her files to the openmm-forcefields repo, and @rafwiewiora is working on doing this now.

In other words, absolutely nothing has been done in the last month.  That's what I was afraid of.  Though it would be nice if you would actually come out and say it.

**Comment by jchodera** [comment]
> In other words, absolutely nothing has been done in the last month. That's what I was afraid of. Though it would be nice if you would actually come out and say it.

Very little effort from my lab has been directed toward this project in the last month because all of our effort was dedicated to more important tasks.

Would you like us to actually focus our effort on it in the next week, or do you want to finish it yourself?

**Comment by jchodera** [comment]
I am the boss, so the blame certainly lies with me. I directed our effort toward more important things, like getting the lab funded and publishing papers. Feel free to blame me if you like, but I'm offering to direct our resources toward this over the next week so we can actually wrap this up.

**Comment by peastman** [comment]
Then let's plan to finish it up in the next week.

**Comment by jchodera** [comment]
Sounds great. @ChayaSt and @rafwiewiora will both be around, and my schedule is clear.

**Comment by jchodera** [comment]
@peastman : We now have conda-installable AmberTools 16 forcefields! Can you test this?

Install the `openmmforcefields` package from the `dev` channel:
```
conda install --yes -c omnia/label/dev openmmforcefields
```
In Python, try creating a `ForceField` object from the new `protein.ff14SB.xml` forcefield:
```
>>> from simtk.openmm import app
>>> ff = app.ForceField('amber/protein.ff14SB.xml')
```

**Comment by peastman** [comment]
Yay!  I'll give it a try.

**Comment by peastman** [comment]
Trying to install gives the following error:

```
PackageNotFoundError: Dependency missing in current osx-64 channels: 
  - openmmforcefields -> openmm >=7.2.0 -> fftw3f
```

Is this Linux only right now?

**Comment by jchodera** [comment]
There are [`osx` packages](https://anaconda.org/omnia/openmmforcefields/files) up right now.

`openmmforcefields` has the requirement `openmm >=7.2.0`, and it may not be recognizing your current installed OpenMM version as satisfying this. We are doing nightly builds of `openmm 7.2.0` from `master` in the `dev` label, and you could install that with
```
# force downgrade to 7.1.1
conda install --yes openmm==7.1.1
# clean cache
conda clean -plti --yes
# force upgrade to 7.2.0
conda install --yes -c omnia/label/dev openmm==7.2.0
```
which should update you to what conda thinks is 7.2.0. 

I imagine you can then install your development version of OpenMM from source over this and it would still continue to think you have 7.2.0.

**Comment by peastman** [comment]
I tried to do something similar to that recently on a different computer.  It led to a corrupt miniconda install, and I eventually had to delete the whole thing and reinstall everything from scratch.  Is there a different way I can get the forcefields?

**Comment by jchodera** [comment]
There are a multitude of ways!

If you're using your own installation of OpenMM from source, you probably want
```
pip install git+git://github.com/choderalab/openmm-forcefields.git
```
It will still check if you have `openmm >=7.2` installed, however. If you run into trouble, can you dump some info about what version of OpenMM conda thinks you have installed?
```
conda list | grep openmm
```

**Comment by peastman** [comment]
Thanks!  That worked.

**Comment by peastman** [comment]
I successfully installed it with conda by creating a new environment, so it could install the conda OpenMM build.

**Comment by jchodera** [comment]
Yay! And it found the forcefields?

**Comment by peastman** [comment]
Yes, I was able to create a ForceField object from one of them.

**Comment by peastman** [comment]
It looks like we have some users who could really use the CHARMM files as soon as they're ready.  See https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=7939&p=0&start=0&view=.

**Comment by jchodera** [comment]
Awesome! It will be useful to have more folks to help test.

**Comment by jchodera** [comment]
The AmberTools17 converted forcefields have been converted and are available in [this directory](https://github.com/choderalab/openmm-forcefields/tree/master/amber/ffxml).

The master index of which LEaP files are converted into `ffxml` files is [here](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml).

Which forcefields should be considered as essential for distribution in OpenMM 7.2?
Note that all these files will be available for conda installation as `openmm-forcefields`, and accessible as `amber/[filename]` via the plugin feature.

@swails @hainm @rafwiewiora : Input would be welcomed!

**Comment by peastman** [comment]
In addition to including some subset of those files, we might also want to create index files that let users use recommended combinations of them (for example, a protein file, a DNA file, and an RNA file) by just including one file.  So there's really two questions.  Which of those files should we bundle with OpenMM?  And what are the recommended combinations we want to make it especially easy to use?

**Comment by peastman** [comment]
I'd really appreciate feedback on this from those of you more knowledgeable about Amber than I am.  Based on looking at the manual, we probably want to provide a file called amber14sb.xml that includes the following:

protein.ff14SB.xml
DNA.OL15.xml
RNA.OL3.xml

Which other files do we want to include?  It sounds to me like all the other DNA, RNA, and protein files are out of date and not recommended (except for possibly protein.ff14SBonlysc.xml?).  We should include gaff2.xml and lipid17_merged.xml (but call it just lipid17.xml).  What about gaff.xml?  And then there's all the water models.  Probably just the standard version of each one?

Keep in mind that all these files will be available through a separate conda package, so we're not limiting people from using any of them.  It's just a question of which are the recommended ones we want to bundle with OpenMM.

**Comment by jchodera** [comment]
`gaff.xml` and `gaff2.xml` won't be usable without additional infrastructure, so it may be best to let us install those with `openmm-forcefields` along with the necessary infrastructure.

**Comment by jchodera** [comment]
Tagging @swails @hainm @rafwiewiora to see if they have additional input!

@peastman : Let me also email David Case (CCing you) to ask for suggestions.

**Comment by hainm** [comment]
@peastman 

> protein.ff14SB.xml

confirmed

> DNA.OL15.xml

I am not sure about this. The previous popular force field for DNA is bsc0 and the it's evolved to bsc1. 
From a quick google search (bsc1 vs OL15), it seems that they are quite competitive. 

> RNA.OL3.xml

likely 

> It sounds to me like all the other DNA, RNA, and protein files are out of date and not recommended (except for possibly protein.ff14SBonlysc.xml?)

only comment about: `protein.ff14SBonlysc.xml`, it's recommended for gb-neck2 model. The another protein.ff14SB (without "onlysc") evolved from 14SBonlysc (a bit tweaking in backbone) to work better with TIP3P model.

@jchodera yes, @dacase would be the best person.

**Comment by peastman** [comment]
>  From a quick google search (bsc1 vs OL15), it seems that they are quite competitive.

At the start of Chapter 3 of the Amber manual, there's a table listing their recommendations for each part of the force field.  OL15 is what it recommends for DNA.

**Comment by hainm** [comment]
ah yeah

<img width="668" alt="screen shot 2017-10-25 at 8 52 55 pm" src="https://user-images.githubusercontent.com/4451957/32030095-b802bec2-b9c6-11e7-8649-494c5c469250.png">

**Comment by peastman** [comment]
For water, I notice you only converted TIP3P, SPC/E, and TIP4P-EW.  How about also converting TIP3P-FB and TIP4P-FB?  Those are probably the very best non-polarizable 3 and 4 point water models available.

**Comment by jchodera** [comment]
> For water, I notice you only converted TIP3P, SPC/E, and TIP4P-EW. How about also converting TIP3P-FB and TIP4P-FB? Those are probably the very best non-polarizable 3 and 4 point water models available.

@rafwiewiora :These must have been added to AmberTools 17. Would it be possible for you to quickly add them to your [`master.yaml` file here](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml) in a PR? Locally, to get AmberTools17, you'll need to
```
conda remove --yes ambermini
conda install --yes ambertools==17 -c http://ambermd.org/downloads/ambertools/conda/
```

**Comment by peastman** [comment]
Based on this discussion and feedback from David Case, I suggest including the following files:

protein.ff14SB.xml
DNA.OL15.xml
DNA.bsc1.xml
RNA.OL3.xml
lipid17_merged.xml (renamed to lipid17.xml)
tip3p_standard.xml (renamed to tip3p.xml)
spce_standard.xml (renamed to spce.xml)
tip4pew_standard.xml (renamed to tip4pew.xml)

I'll also include TIP3P-FB and TIP4P-FB, once you get those converted.  David Case also suggested including the leaprc.protein.ff15ipq parameters.  I see you have those commented out under the heading, "FFs that couldn't be converted yet due to problems".  What are the issues with those?

I also suggest converting all filenames to lower case, to be consistent with the existing ones.  All these files will go in an "amber" subdirectory, so you would specify them as `'amber/protein.ff14sb.xml'`.  At the top level, we'll provide the file `amber14sb.xml`, which includes the recommended protein, DNA, and RNA files.

I don't see any implicit solvent force fields here.  Is that intentional?

**Comment by hainm** [comment]
> I don't see any implicit solvent force fields here. Is that intentional?

yes, that's intentional. There is no such force field for implicit solvent. Normally people to combine different solvent and force field to look for error cancellation. The better way is to improve the accuracy of both solvent and FF themselves. (Basically we recommend amber users to use gbneck2 with ff14SBonysc).

**Comment by jchodera** [comment]
> I don't see any implicit solvent force fields here. Is that intentional?

We also don't have a way of assigning implicit solvent parameters using the new scheme where we don't have a unique "atom type" for each atom of every residue. Because we couldn't come to a consensus in #560, instead of extending `ffxml` to include more generic assignment logic, it makes sense to just use SMIRNOFF, since it is trivial to code up various GBSA models using SMARTS.

For the nearer term, the plan in https://github.com/choderalab/openmm-forcefields/issues/2 was to modify [`customgbforces.py`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/internal/customgbforces.py) to assign default parameters programmatically if requested since @swails already put in all the logic to *check* if optimal parameters are set. We could do this for 7.2 or punt until later.

**Comment by jchodera** [comment]
> I'll also include TIP3P-FB and TIP4P-FB, once you get those converted. David Case also suggested including the leaprc.protein.ff15ipq parameters. I see you have those commented out under the heading, "FFs that couldn't be converted yet due to problems". What are the issues with those?

@rafwiewiora is working on all of those right now.

For the `ff15ipq` forcefield, there were errors in the forcefield files that were transparently processed without warnings by LEaP but led to different energies when run through ParmEd. David Cerutti said he'd fix those, but we forgot to check when we updated to AmberTools17. @rafwiewiora is checking now.

**Comment by jchodera** [comment]
> I also suggest converting all filenames to lower case, to be consistent with the existing ones. All these files will go in an "amber" subdirectory, so you would specify them as 'amber/protein.ff14sb.xml'. At the top level, we'll provide the file amber14sb.xml, which includes the recommended protein, DNA, and RNA files.

Our conversion scripts derives the names for `leaprc` conversions from the `leaprc.*` filename. See [the `master.yaml` file](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml). You should feel free to change the names when you copy them into OpenMM, but it would only cause confusion for those users trying to use our conversions directly if we did the same: If the user wants `leaprc.RNA.OL3`, they know that they can install `openmm-forcefields` and ask for `amber/RNA.OL3.xml` and get exactly the conversion of that file. Similarly, the AMBER user would normally use `leaprc.protein.ff14SB` in tleap, and this gets converted to `protein.ff14SB.xml`.

This kind of flexibility was what we hoped to enable by having you copy what *you* wanted, in the way you wanted to, over to OpenMM: We can be strict so that people familiar with LEaP know where to find the OpenMM equivalents, and you can do whatever you think is best for OpenMM users by copying whatever you want however you want. 

This also avoids unnecessary fistfights. :)

**Comment by peastman** [comment]
I tried to combine the protein, DNA, and RNA files, but it fails with an exception:

```
ValueError: Found multiple definitions for atom type: C
```

The problem is that each file has duplicate definitions for the atom types.  Since atom type C gets used in proteins, in DNA, and in RNA, it gets defined three separate times.  I think we need to split out all the atom type definitions into their own file.

**Comment by peastman** [comment]
Alternatively, we could have three different atom types: `protein-C`, `DNA-C`, and `RNA-C`.  It's not just the atom types that are getting duplicated.  Every bonded and nonbonded term also appears multiple times.

**Comment by swails** [comment]
That is a result of the fact that the protein and nucleic acid force fields were once in the same files and shared parameters. 

The separation approach used was to simply put the shared parameters in both parameter files. 

Can ForceField be modified not to throw any error if the “multiple” parameters are equal?

**Comment by jchodera** [comment]
> Can ForceField be modified not to throw any error if the “multiple” parameters are equal?

How would we be able to check if all parameters involving a redefined atom type are equal?

**Comment by jchodera** [comment]
@swails: How does LEaP treat redefined parameters?

**Comment by swails** [comment]
Last one loaded “wins”

**Comment by jchodera** [comment]
What if we had a top-level ffxml `<ForceField>` tag property like `override='amber'` which provides the "last one loaded wins" behavior?

**Comment by jchodera** [comment]
I have a feeling that there are going to be problems if we use a unique prefix for each file. That's why we added the override level support in the first place, though the integer priority really doesn't work here.

**Comment by peastman** [comment]
If you have multiple `<Bond>`, `<Angle>`, etc. tags for the same atom types, we add multiple bonds.  There's no concept of overriding.

**Comment by jchodera** [comment]
What are the implications for adding prefixes to atom types? I think we had mainly wanted to avoid doing this for small molecules, but are there cases where the user would want to add another `ffxml` file that defines additional protein residues but might not have appended the same prefix? How many prefixes do we need? One per file, or one per class of biopolymer?

**Comment by peastman** [comment]
> How many prefixes do we need? One per file, or one per class of biopolymer?

If two files might reasonably be included together, the atom type names need to be unique.  If two files should never be included together, it's OK for the names to collide.  In fact, it could be good, since it will produce an exception.  So if you used the name `DNA-C` and the user included both the OL15 and bsc1 versions, it would (appropriately) throw an exception.

**Comment by rafwiewiora** [comment]
TIP3P-FB and TIP4P-FB ready: https://github.com/choderalab/openmm-forcefields/pull/43, working on the ipq.

**Comment by jchodera** [comment]
OK, so the prefixes `protein-`, `DNA-`, and `RNA-` should be sufficient? 
I'll extend the conversion script to allow us to specify an atom type prefix to be applied to every non-wildtype type in the parameter set. I'm just trying to figure out whether the code for this should be within ParmEd (at reading time? a method in the parameter set? at writing time?) or within the conversion script (which would iterate through the parameter set and append prefixes to everything).

**Comment by peastman** [comment]
Sounds good.  You can specify `name="DNA-C" class="C"`.  That way if someone wants to add another file that refers to the standard types, they'll have the option of using `class` to refer to all of them without having to distinguish between prefixes.

**Comment by dacase** [comment]
On Mon, Oct 30, 2017, John Chodera wrote:

> OK, so the prefixes `protein-`, `DNA-`, and `RNA-` should be sufficient? 
> I'll extend the conversion script to allow us to specify an atom type
> prefix to be applied to every non-wildtype type in the parameter
> set. I'm just trying to figure out whether the code for this should
> be within ParmEd (at reading time? a method in the parameter set? at
> writing time?) or within the conversion script (which would iterate
> through the parameter set and append prefixes to everything).

Just a note of Amber history here: when we split up the parameter files in the
Spring of 2016 (for the AmberTools16 release), we did pretty extensive testing
that the results were identical to what one obtained with the "combined" files
that were used earlier.  The results were pretty convincing that atom types
used in common for proteins and nucleic acids really were identical.

Of course, it's possible that we missed something then, or that things have
diverged in the 18 months since then (but I don't think so.)  I'm a little
concerned if one adopts a prefix scheme, that might either (a) cause people
to think a prefix is really needed; or (b) encourage people to create a
divergence that would be hard to fold back into Amber itself (admittedly our
problem not yours.)

I'm not sure what the right answer is here.  If an openMM/ParmEd approach
could be used to get closer to program-independent force-field files, the
Amber developers should support that.  But getting there may be hard.  

I wonder if there is not some way (as an intermediate step, and a special
case) to have the workflows you are creating be smart enough to know that
atom types for Amber's existing protein, DNA and RNA force fields are
in fact consistent?  That is, a duplicate atom type definition would be
allowed if it fell into a specific class that would be grandfather-ed in;
otherwise some sort of error condition could be triggered.

Apologies if I am missing some critical detail here.

...thx...dac

**Comment by peastman** [comment]
In order to do that, we would need to put everything common into a single file.  The "proteins" file would contain only definitions specific to proteins and the DNA file would contain only definitions specific to DNA.  Anything common to them would be in its own file.

**Comment by dacase** [comment]
re: put everything common into a single file:  That is a good idea, and probably what Amber should do going forward.  It's simple for us to implement, and there's a clear advantage if file contents and file naming schemes can be in sync between Amber and your efforts.

**Comment by peastman** [comment]
@jchodera what's the status of resolving the remaining issues?

**Comment by jchodera** [comment]
I’m at a Simons Foundation (speaking) all day today but will update this thread tonight!

**Comment by peastman** [comment]
So, what's the status?

**Comment by jchodera** [comment]
So sorry for the delay here, @peastman---Simons let out late Tue and I've been stuck in meetings all day. 

Here's the current status, which reflects the updated [project board](https://github.com/choderalab/openmm-forcefields/projects/1):

(Note the first item, which could use your input!)

### AMBER
* See https://github.com/choderalab/openmm-forcefields/issues/46 : We have to decide whether to merge `protein.ff14SB.xml`, `DNA.OL15.xml`, and `RNA.OL3.xml` together into a single file (merging atom types that are identical, as [@dacase suggests](https://github.com/pandegroup/openmm/issues/1611#issuecomment-340848977)) or modify ParmEd to append prefixes to all atom types. The ParmEd modifications would likely take several days to complete if we decided to add prefixes; combination of these into one file can likely be performed immediately.
* See https://github.com/choderalab/openmm-forcefields/issues/44 : We had chosen to punt on `parm15ipq`, the last remaining recommended forcefield to convert that appeared to have torsion bugs in the LEaP input representation, but @dscerutti has suggested [a fix](https://github.com/choderalab/openmm-forcefields/issues/44#issuecomment-343880821) that @rafwiewiora will try out ASAP. If it works, we'll include it; if not, we'll punt.

### CHARMM
* See https://github.com/choderalab/openmm-forcefields/issues/34 : Conversion of non-CHARMM water models is mostly complete, but is currently blocked by waiting for @swails to merge https://github.com/ParmEd/ParmEd/pull/914 to allow the final file to be converted
* See https://github.com/choderalab/openmm-forcefields/issues/22 : Patch debugging is awaiting me to finish https://github.com/ParmEd/ParmEd/pull/895 which is essentially complete but was causing some weird travis errors that have hopefully been resolved
* See https://github.com/choderalab/openmm-forcefields/issues/15 : I need to automate the CHARMM energy tests

Anticipated ETA for my bits: I have no meetings Nov 23-28 so should get everything wrapped up then at the very latest, but I'll be able to push on anything I'm blocking on this weekend.

**Comment by dacase** [comment]
About Amber: I think my comment 16 days ago (that John refers to) has been misinterpreted: I was
agreeing with Peter's suggestion right before mine: that we actually have four files: protein.ff14SB.xml, DNA.OL15.xml, RNA.OL3.xml, and a new file, "Amber_common.xml" that contains anything duplicated in the first three files.  (Those duplicates would then be removed from the first three files.)

Main advantage is that the file naming scheme would look the same as in Amber (even more so after we release AmberTools18 and go to the "common file" idea ourselves.  Also, Amber users are certainly going to want to (for example) swap in the bsc1 DNA force field in place of DNA.OL15; that would be nearly impossible if you re-merge everything into a single .xml file.  Finally, it will really complicate the process of keeping current with Amber going forward, since we are pretty committed to having separate files for proteins, DNA and RNA.

Of course, I'm the outsider here.  But from my point of view, either using this "common file" to hold what would otherwise be duplicates, or using ParmEd-generated  prefixes, both seem preferably to re-merging the protein, DNA and RNA .xml files.

**Comment by jchodera** [comment]
Thanks for the clarification, @dacase! This does sound logical to me.

Regarding swapping in the [`DNA.bsc1.xml`](https://github.com/choderalab/openmm-forcefields/blob/master/amber/ffxml/DNA.bsc1.xml) DNA forcefield (or other replacements for the defaults): 
1. Presumably I'd also have to strip out any parameters in `Amber_common.xml` from [`DNA.bsc1.xml`](https://github.com/choderalab/openmm-forcefields/blob/master/amber/ffxml/DNA.bsc1.xml) in order to make this work
2. If any force terms (atom type definitions, bonds, angles, dihedrals, impropers, or LJ definitions) share a common atom name but differ in any parameter, this process would fail and we would be unable to include that forcefield in this scheme until the source files were fixed.

@peastman: Would you be able to help with this? I fear this change would push back my own timeline by at least a month, if not more. As far as I can tell, this requires:
* Implementing set operations on forcefield parameter sets into ParmEd, especially `intersect` (for determining the contents of `Amber_common.xml`) and `difference` (for removing the contents of `Amber_common.xml` from the specific forcefields like `DNA.OL15.xml`), along with corresponding unit tests.
* Modifying the [Amber conversion script](https://github.com/choderalab/openmm-forcefields/blob/master/amber/convert_amber.py) and corresponding [master YAML file](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml) to specify how `Amber_common.xml` should be constructed and how its contents should be subtracted from the other forcefields
* Testing the conversion script to ensure there are no duplicate atom names with different parameters in any force term, and coordinating with Amber forcefield developers to reconcile those differences

The good news is that, because ParmEd forcefield representations are forcefield-agnostic, the new features added to ParmEd could help the Amber devs do the same thing for the Amber-format forcefields for the next AmberTools18, though they may already have tools for that (which we may want to use instead if they already exist?).

**Comment by peastman** [comment]
> Testing the conversion script to ensure there are no duplicate atom names with different parameters in any force term, and coordinating with Amber forcefield developers to reconcile those differences

I can check that directly from the original Amber data files.  I'll do that.

From a quick look at them, it appears that part (maybe all) of the problem is in the way you're doing the conversion rather than how Amber defines the force fields.  For example, when I look at your converted `DNA.OL15.xml` file, I see the `<HarmonicBondForce>` section defines 75 sets of bond parameters.  But when I look at `frcmod.DNA.OL15`, it only has 18.  All the others are (I believe) coming from `parm10.dat`, which you're including with each of the different files.  That's why you have so many duplicate definitions.

**Comment by jchodera** [comment]
Thanks for looking into this, @peastman.

The [`master.yaml`](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml) file that directs the conversion simply uses the contents of the `leaprc` files for most of the conversions. It's these `leaprc` files provided by AMBER that seek to load the `parm10.dat` parameter set, which appear to be essential since the ff14SB modifications are delivered as an `frcmod` file that only modifies the base parameters. For example, in `leaprc.protein.ff14SB`, we have:
```
#
#       Load the main parameter set.
#
parm10 = loadamberparams parm10.dat
frcmod14SB = loadamberparams frcmod.ff14SB
```
Our goal is to avoid modifying the `prmtop` files that Amber has carefully compiled, and even if we did separate out `parm10.dat` to a separate `ffxml` file, my understanding is that the `frcmod.ff14SB` file still *modifies* some of the  parameters defined there (e.g. torsions) which is not permitted by `ForceField`.

**Comment by peastman** [comment]
Here's the corresponding section from `leaprc.DNA.OL15`:

```
parm10 = loadamberparams parm10.dat
loadamberparams frcmod.DNA.OL15
```

So does that mean the parameter modifications in `frcmod.ff14SB` get applied to proteins but not to DNA?  If so, then applying prefixes is the only real option because types with the same name in different files actually have different parameters!  Perhaps you're relying on those types only ever getting used by one type of molecule and not the other?  But it's impossible to guarantee that because some definitions involve wildcards.

**Comment by rafwiewiora** [comment]
I think if you're looking to waste a lot of time doing this `Amber_common.xml` thing is the way to go... I say let the AMBER folks do the switch for the next year release and that we should be able to convert without any problems on our side. 

For now I think prefixes are the best. It won't take a few days @jchodera - literally all you need to do is copy 5 lines starting here: https://github.com/choderalab/openmm-forcefields/blob/master/amber/convert_amber.py#L222 into `convert_leaprc` - I can get this done today together with the `15ipq`.

**Comment by peastman** [comment]
Sounds fantastic.  Thanks!

**Comment by jchodera** [comment]
@rafwiewiora : Does the code in [these lines](https://github.com/choderalab/openmm-forcefields/blob/master/amber/convert_amber.py#L222-L226) also ensure that any bonds, angles, and torsions (including torsions with wildcards) contains the modified atom names too?

**Comment by rafwiewiora** [comment]
Hmm - I'll tell you as soon as I test it.

**Comment by jchodera** [comment]
I'm pretty sure it doesn't. Instead of pointers to `AtomType` objects, it looks like all of the bonds, angles, torsions, impropers, etc. in [`ParameterSet`](https://github.com/ParmEd/ParmEd/blob/master/parmed/parameters.py#L21-L73) use strings to refer to atom types. I still think it will take quite some time to add the ability to robustly add a prefix to a specific atom type name in a `ParameterSet` and have ParmEd continue to work correctly. For me to make this change, it would be at least a week to implement and debug, assuming nothing major went wrong.

**Comment by jchodera** [comment]
We should probably post an issue on the [ParmEd issue tracker](https://github.com/ParmEd/ParmEd/issues) to see if we can get any insight from @swails

**Comment by rafwiewiora** [comment]
No need to modify `ParameterSet` - let's just add the prefix at writing to `.xml` - give these functions: https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L230-L242 a `prefix` argument?

**Comment by peastman** [comment]
Or let's just modify the existing files to add a prefix.  Here, I just wrote a script for you.

```Python
import re
import sys

prefix = sys.argv[2]
inTypes = False
replacements = {}

for line in open(sys.argv[1]):
    if '<AtomTypes>' in line:
        inTypes = True
    if '</AtomTypes>' in line:
        inTypes = False
    if inTypes:
        match = re.search('name="(.*?)"', line)
        if match is not None:
            name = match.group(1)
            newName = prefix+name
            line = line.replace('name="%s"' % name, 'name="%s"' % newName)
            replacements['type="%s"' % name] = 'type="%s"' % newName
            replacements['type1="%s"' % name] = 'type1="%s"' % newName
            replacements['type2="%s"' % name] = 'type2="%s"' % newName
            replacements['type3="%s"' % name] = 'type3="%s"' % newName
            replacements['type4="%s"' % name] = 'type4="%s"' % newName
    else:
        for key in replacements:
            if key in line:
                line = line.replace(key, replacements[key])
    if line.endswith('\n'):
        line = line[:-1]
    print(line)
```

You can invoke it as

```
python addPrefix.py DNA.OL15.xml DNA-
```

**Comment by jchodera** [comment]
Would that work for wildcards?

**Comment by peastman** [comment]
Yes.  Why wouldn't it?  Wildcards are specified by `""`, and those are left unchanged.

**Comment by jchodera** [comment]
OK, with this hacky workaround (which will still require writing more tests on my end) and the merging of https://github.com/ParmEd/ParmEd/pull/914, I think I can make a good deal of headway this weekend! We can do some testing early next week and I can definitively wrap up all remaining TODOs Thanksgiving. 

We seem to have addressed the CUDA 9 `dev` build and CPU performance issues in https://github.com/pandegroup/openmm/issues/1869 as well. 

@peastman : Do you want to open a thread to list any remaining issues / bugfixes we want to address before 7.2 gets closed out?

**Comment by jchodera** [comment]
@peastman : Just want to make sure you see this question regarding duplicate residue names: https://github.com/choderalab/openmm-forcefields/issues/16#issuecomment-345534553

**Comment by jchodera** [comment]
@peastman : I've built a [`charmm36.xml` metapackage](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/ffxml/charmm36.xml) and converted [CHARMM36 and water/ion files separately now](https://github.com/choderalab/openmm-forcefields/tree/master/charmm/ffxml). I am no longer seeing duplicate residue errors, but your POPC example is failing with a cryptic error message:
```python
lski2414:charmm choderaj$ python
Python 3.6.1 |Continuum Analytics, Inc.| (default, May 11 2017, 13:04:09) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from simtk.openmm import app
>>> pdbfile = app.PDBFile('tests/POPC.pdb')
>>> ff = app.ForceField('ffxml/charmm36.xml')
>>> system = ff.createSystem(pdbfile.topology)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py", line 1270, in createSystem
    force.createForce(sys, data, nonbondedMethod, nonbondedCutoff, args)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py", line 2388, in createForce
    reverseMap[typeMap[typeValue]] = typeValue
IndexError: list assignment index out of range
```
I'm not quite sure what's causing this failure, but it also took several minutes for the `ff.createSystem()` line to run.

I've attached your PDB file:
[POPC.pdb.zip](https://github.com/pandegroup/openmm/files/1485941/POPC.pdb.zip)

**Comment by jchodera** [comment]
@peastman : Here's a summary of where we stand after this weekend.

Current ffxml files are here (note that you have write access to this repo):
https://github.com/choderalab/openmm-forcefields

Project board (to track status of outstanding tasks) is here:
https://github.com/choderalab/openmm-forcefields/projects/1

## AMBER
* AMBER ffxml files have been updated with prefixes and final ffxml files are individually tested with protein, lipids, and nucleic acids (https://github.com/choderalab/openmm-forcefields/pull/51)
* The lipid merging has been incorporated into the AMBER conversion and testing script in the PR  https://github.com/choderalab/openmm-forcefields/pull/52 but tests are currently failing because of a bond discrepancy. It would be great to get some help with debugging this failure.

## CHARMM
* CHARMM ffxml files now have water models separated out (https://github.com/choderalab/openmm-forcefields/pull/42)
* A [`charmm36.xml` metapackage](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/ffxml/charmm36.xml) has been built to bring in the default water model (https://github.com/choderalab/openmm-forcefields/pull/42)
* CHARMM ffxml tests are now automated (https://github.com/choderalab/openmm-forcefields/pull/48) but failing due to energy discrepancies. I could use your help chasing this down since I won't be able to work on this again until Friday (where I'll have a few days to finish any outstanding work).
* Parameterizing your `POPC.pdb` fails with [an error message](https://github.com/choderalab/openmm-forcefields/issues/22#issuecomment-345546220) and takes several minutes, so any help on speeding this up or fixing the failure would be very much appreciated.

**Comment by peastman** [comment]
Are you running an old version of the code?  I think that error was fixed in #1915.

I just tried using the same files, and I get this error instead:

```
ValueError: No template found for residue 1 (POP).  The set of atoms matches POPC, but the bonds are different.  Perhaps the chain is missing a terminal group?
```

I'll take a look to see why it's not matching.

**Comment by jchodera** [comment]
Sorry for not having been explicit here. That error is because the AMBER tool is expecting the residue name `POPC`, but your file had `POP`. In https://github.com/choderalab/openmm-forcefields/pull/52#issuecomment-345589975, I said
> The AMBER conversion tool first doesn't understand your residue names, and even if they are modified to match, it doesn't understand your atom names to convert from CHARMM to AMBER.

but I should have explicitly stated that I had tried replacing the residue name `POP` to `POPC` (or you could edit the CSV file) but then ran into trouble with your atom names not matching CHARMM atom names.

**Comment by peastman** [comment]
It's impossible to create a PDB file with a residue named POPC.  The PDB file format only allows three columns for the residue name!

It shouldn't matter though.  As long as the PDB file has correct bonds and elements, we should be able to match it.

**Comment by peastman** [comment]
I found the problem with the PDB files and updated them:

https://raw.githubusercontent.com/peastman/openmm/membrane/wrappers/python/simtk/openmm/app/data/POPC.pdb
https://raw.githubusercontent.com/peastman/openmm/membrane/wrappers/python/simtk/openmm/app/data/POPE.pdb

I verified the above code now works correctly.

**Comment by jchodera** [comment]
> I verified the above code now works correctly.

Which code? Calling `charmmlipid2amber.py` on those PDB files? If so, I can update the AMBER conversion tests with these files.

**Comment by peastman** [comment]
I'm referring to the code in this comment: https://github.com/pandegroup/openmm/issues/1611#issuecomment-345546490

**Comment by jchodera** [comment]
Ah! `System` creation using `charmm36.xml` to `POPC.pdb`!

**Comment by jchodera** [comment]
I'm still seeing the same failure:
```
lski2414:charmm choderaj$ python
Python 3.6.1 |Continuum Analytics, Inc.| (default, May 11 2017, 13:04:09) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from simtk.openmm import app
>>> ff = app.ForceField('ffxml/charmm36.xml')
>>> pdbfile = app.PDBFile('POPC.pdb')
>>> ff.createSystem(pdbfile.topology)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py", line 1270, in createSystem
    force.createForce(sys, data, nonbondedMethod, nonbondedCutoff, args)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py", line 2388, in createForce
    reverseMap[typeMap[typeValue]] = typeValue
IndexError: list assignment index out of range
```
Also, it still takes a couple of minutes---do you find the same?

**Comment by peastman** [comment]
You're using an old version of the code.  That line doesn't exist in the latest version.

**Comment by jchodera** [comment]
Let's be clear what we mean by "the code" here since there are at least four in play (OpenMM, openmm-forcefields, ParmEd, AMBER). Do you mean I should update to the latest OpenMM dev build?

**Comment by jchodera** [comment]
Hm, OpenMM `dev` conda builds broke six days ago:
<img width="1257" alt="broken-dev-builds" src="https://user-images.githubusercontent.com/3656088/33045482-efe36e28-ce1a-11e7-8145-4db6cfa923b1.png">

**Comment by jchodera** [comment]
Looks like the [`dev` build failure might be a `conda-build` issue](https://travis-ci.org/omnia-md/conda-dev-recipes/jobs/304822474#L7275-L7294). I'll have @Lnaden look into that in the morning.

**Comment by peastman** [comment]
You're using code that's much older than that.  This was changed in #1915, which was merged on Oct. 7.

**Comment by jchodera** [comment]
I've just updated to the latest openmm `dev` build, `7.2.0.dev-cee3b81`.

It *does* succeed now, but it takes 76 seconds on my laptop to build a system that only has 32K atoms. Anything that can be done to speed this up?

**Comment by peastman** [comment]
I'll take a look at the speed in a little bit and see what's going on.

**Comment by Lnaden** [comment]
Found the problem. Not sure who is in charge of it, but we got this warning on one of the last builds which passed:
`/opt/conda/lib/python3.6/site-packages/binstar_client/scripts/cli.py:72: UserWarning: Token warning: expires in 2 days`
Seen here:
https://travis-ci.org/omnia-md/conda-dev-recipes/jobs/300875567#L7219

So our Binstar token expired. Time to get a new one. Must of been decently old to still be called a "Binstar token"

**Comment by jchodera** [comment]
I can fix that quickly.

**Comment by peastman** [comment]
I looked into the slowness.  Most of the time is spent on the proper torsions.  It seems to just be that this is a ridiculously large force field.  It contains 6164 different proper torsion definitions, compared to 203 in protein.ff14SB.xml.  The complete file charmm36_nowaters.xml is 112,381 lines, compared to 3441 for protein.ff14SB.xml.

**Comment by jchodera** [comment]
I guess our options are:
1. Tolerate the slowness
2. Split up CHARMM36 into multiple components (which we decided not to do)
3. Use a clever data structure to more rapidly assign torsions

**Comment by jchodera** [comment]
@peastman : I'm mostly out of commission until Thursday, but I've updated the [CHARMM PR](https://github.com/choderalab/openmm-forcefields/pull/48) and [AMBER PR](https://github.com/choderalab/openmm-forcefields/pull/52) with questions that could benefit from your investigation.

**Comment by jchodera** [comment]
I can't join the OpenMM call today, but am working on the following:
* [AMBER](https://github.com/choderalab/openmm-forcefields/pull/52): Debugging [bond issues](https://github.com/choderalab/openmm-forcefields/pull/52#issuecomment-346437889) in lipid merging code followed by the addition of tests that ensure protein + solvent + nucleic acids + lipids can be included together without issues.
* [CHARMM](https://github.com/choderalab/openmm-forcefields/pull/48): Splitting out all [model compounds](https://github.com/choderalab/openmm-forcefields/pull/48#issuecomment-346422431) followed by the addition of more test systems generated by CHARMM-GUI (vacuum protein, protein+small molecule, protein+membrane)

Other than these issues (which are the only remaining ones on the [project board](https://github.com/choderalab/openmm-forcefields/projects/1)) there isn't anything standing in the way of the forcefield conversion from being complete.

**Comment by peastman** [comment]
Currently `amber14sb.xml` includes files for proteins, DNA, and RNA.  Should it also include lipids?

**Comment by jchodera** [comment]
Which file are you talking about? Something in the OpenMM distribution, or a file from [`openmm-forcefields`](https://github.com/choderalab/openmm-forcefields/tree/master/amber/ffxml)?

**Comment by peastman** [comment]
The index file included in OpenMM.  See https://github.com/pandegroup/openmm/issues/1611#issuecomment-339464070.

**Comment by jchodera** [comment]
Ah! Sure, why not?

**Comment by dacase** [comment]
On Mon, Nov 27, 2017, peastman wrote:

> Currently `amber14sb.xml` includes files for proteins, DNA, and RNA.
> Should it also include lipids?

Possible problem in confusing Amber users.  Currently, "amber14sb.xml" has
the same information as the *former* (before AmberTools16) "leaprc.ff14sb"
file.  Most Amber users will recognize this.

Adding lipids to the same file never happened with Amber, so that could
be confusing.  Also, the moniker "14sb" (which correctly only refers to the
protein parameters) would become even less appropriate.

If lipids get added, maybe the file name should be something like
"amberxx.xml" (without the "sb"), presumably with comments inside, or other
documentation" about where the parameters came from.  Not sure what "xx"
should be, but it's worth thinking about how a future updates for all these
.xml files will be named.

....dac

**Comment by jchodera** [comment]
Note that we *do* have a direct conversion of most of the `leap.*` files (indexed [here](https://github.com/choderalab/openmm-forcefields/blob/master/amber/files/master.yaml)) that will be available to users as `amber/ff14sb.xml` or `amber/leaprc/ff14sb.xml` or similar. But @peastman's bundles will be available at the top level---just `filename.xml`.

For bundles that deviate from Amber's naming convention, I do agree that it would be best to use a non-confusing naming convention.  Maybe we can make this explicit by using names like `amber-bundled-14.xml` or `amber14-bundled.xml`?

The converted XML files do contain references and provenance information, but we haven't discussed whether the bundles @peastman is rolling should as well. What's the best way to approach this?
1. Include an `<Info/>` block in these bundles so they're easy to inspect
2. Have `app.ForceField` expose the metadata so that API users can easily know what is in the files
3. Focus on the User Guide and list which Amber files went into each bundle

**Comment by peastman** [comment]
@dacase What name do you think would best convey to Amber users that the file contains the recommended parameters for all molecule types from this particular family of force fields?  Would `amber14.xml` or perhaps `amber14-all.xml` be clear?  Is there a name people typically use to refer to the whole family rather than just the protein parameters?

On a different subject, I'm now looking at the CHARMM water models and trying to figure out which ones to include, what to call them, and how to document them.  I'm a bit confused about the file `water_ions_default.xml`.  Looking at the parameters, I see it includes a LJ term on the hydrogens, so this is the modified CHARMM version of TIP3P, not the regular one.  However,

1. The reference contained in the file is to a paper that discusses standard TIP3P, not the modified one.
2. According to http://pubs.acs.org/doi/abs/10.1021/acs.jpcb.6b01316, the CHARMM36 protein parameters were parametrized with standard TIP3P, *not* the modified one, and it produces bad results when used with the modified one.

Most of the other files have no LJ term on the hydrogens, as you would expect.  But there are two exceptions: SPC and SPC/E.  Those specify

```xml
<Atom type="HT" sigma="1.7817974362806786e-05" epsilon="4.184e-06"/>
```

Epsilon is small enough that this probably won't have much effect, but even so, that is *not* the definition of SPC or SPC/E.

**Comment by dacase** [comment]
On Mon, Nov 27, 2017, peastman wrote:

> @dacase What name do you think would best convey to Amber users that the
> file contains the recommended parameters for all molecule types from
> this particular family of force fields?  Would `amber14.xml` or perhaps
> `amber14-all.xml` be clear?  Is there a name people typically use to
> refer to the whole family rather than just the protein parameters?

There is no name that Amber people are used to.  I'd think that
"amber14-all.xml" would be a good choice.

....dac

**Comment by peastman** [comment]
Ok, I'll go with that.  Thanks!

**Comment by jchodera** [comment]
> On a different subject, I'm now looking at the CHARMM water models and trying to figure out which ones to include, what to call them, and how to document them. I'm a bit confused about the file water_ions_default.xml. Looking at the parameters, I see it includes a LJ term on the hydrogens, so this is the modified CHARMM version of TIP3P, not the regular one.

It's always good to go back to the source for questions like this. This conversion is directed by [`waters.yaml`](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/files/waters.yaml#L3-L24), and @ChayaSt has copied over the references appearing in the [CHARMM source file `toppar/toppar_water_ions.str`](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/toppar_water_ions.str). You'll see the Jorgensen reference appears there as well.

If there are mistakes (and there are plenty!) we should be contacting Alex MacKerell to lobby him to update the official CHARMM36 distribution to correct these annotation errors. As curators, however, I think it's important we try to represent whatever those files contain as closely as we can, even if the information is wrong.

These are, however, the recommended standard CHARMM36 water parameters: See the [README](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/00toppar_file_format.txt#L199-L214), which contains an example that loads these files. They're also the only water files not shoved into another directory.

I'd suggest you email Alex and CC me on those two points. I do think they'd be important to correct (or at least clarify).

**Comment by jchodera** [comment]
> Most of the other files have no LJ term on the hydrogens, as you would expect. But there are two exceptions: SPC and SPC/E. Those specify...

The [README](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/00toppar_file_format.txt#L156-L167) also discusses this:
> (F) Additional topology and parameter files from various sources are included in subdirectories of the toppar directory.  A description follows:
> subdirectory non_charmm: Contains toppar files for AMBER, Bristol-Myers Squibb (BMS) and OPLS force fields along with stream files for a variety of water models.  These files have been tested to the extent that they may be considered reliable representations of the original force fields, though potentially not exact representations.  These files are NOT maintained and, thus, use at your own risk.  See the 00readme files and note that AMBER requires a special version of CHARMM as described in the 00readme file.

TL;DR: "We don't claim anything is correct in there; use at your own risk."

Again, it would be useful to let Alex know about these issues.

**Comment by peastman** [comment]
Are there any remaining issues?  Or are the files at

https://github.com/choderalab/openmm-forcefields/tree/master/amber/ffxml
https://github.com/choderalab/openmm-forcefields/tree/master/charmm/ffxml

now the final versions, ready to be copied to OpenMM?

**Comment by jchodera** [comment]
There are no known outstanding issues. These files are ready for testing in OpenMM. In parallel, I am working on more automated testing for more complex systems, but I think we should do that in parallel.

The dev builds are all ready to go, so we can test with that before cutting release candidates.

**Comment by peastman** [comment]
Ok, great!  I'll get to work copying them over.

**Comment by rafwiewiora** [comment]
@jchodera just one little thing - you're missing references for the 15ipq: https://github.com/choderalab/openmm-forcefields/blob/master/amber/ffxml/protein.ff15ipq.xml#L5 and https://github.com/choderalab/openmm-forcefields/blob/master/amber/ffxml/protein.ff15ipq-vac.xml

**Comment by jchodera** [comment]
Can you add those ASAP, @rafwiewiora?

**Comment by jchodera** [comment]
Sorry, I hadn't realized you just put placeholders there.

**Comment by rafwiewiora** [comment]
15ipq reference added (https://github.com/choderalab/openmm-forcefields/pull/53)

**Comment by jchodera** [comment]
Closing this now. Please reopen or open a new issue if any problems are noted!
